python/dgl/nn/pytorch/network_emb.py

```diff
@@ -0,0 +1,220 @@
+        metapath are commonly the same.
+    window_size :int
+        The context size which is considered for positive samples.
+    min_count :int , optional
+        The number of node counts that occur less than  min_count  value in all meta-paths
+        will be discarded before random walks. (default: 0)
+    negative_samples_size :int , optional
+        The number of negative samples need to be extracted for each positive sample.(default: 5)
+    node_repeat  :int , optional
+        The number of random walks to sample for each start node$V_1$. (default: 1)
+    nid2word  :Dict[ str, Dict[ int , str ] ], optional
+        If set, we can use model.id2word  to get the dict where the key is embedding id and
+        the value is its corresponding name.(default: None)
+    sparse  :int , optional
+        If set to True, gradients w.r.t. to the weight matrix will be sparse. (default: True)
+    """
+
+    def __init__(self,
+                 g,
+                 emb_dimension,
+                 metapath,
+                 window_size,
+                 min_count=0,
+                 negative_samples_size=5,
+                 node_repeat=1,
+                 nid2word=None,
+                 sparse=True
+                 ):
+        super(Metapath2vec, self).__init__()
+        self.g = g
+        self.emb_dimension = emb_dimension
+        self.metapath = metapath
+        self.min_count = min_count
+        self.window_size = window_size
+        self.negative_samples_size = negative_samples_size
+        self.nid2word = nid2word
+        self.sparse = sparse
+        self.node_repeat = node_repeat
+        self.token_count = 0
+        self.walk_dataset = []
+        self.negatives = []
+        self.discards = dict()
+        self.negpos = 0
+        self.NEGATIVE_TABLE_SIZE = 1e8
+        self._extract()
+        self._initTableNegatives()
+
+    def _extract(self):
+
+        assert len(self.metapath) + 1 >= self.window_size
+        word_frequency = dict()
+        word_frequency_filtered = dict()
+
+        ##build whole vocab for all nodes
+        self.token_count = 0
+        self.type2id = dict()
+        self.id2word = dict()
+        wid = 0
+        if self.nid2word is None:
+            for nodetype in self.g.ntypes:
+                self.type2id[nodetype] = dict()
+                for idx in range(self.g.num_nodes(nodetype)):
+                    self.type2id[nodetype][idx] = wid
+                    wid += 1
+        else:
+            for nodetype in self.g.ntypes:
+                self.type2id[nodetype] = dict()
+                for idx in range(self.g.num_nodes(nodetype)):
+                    self.type2id[nodetype][idx] = wid
+                    self.id2word[wid] = self.nid2word[nodetype][idx]
+                    wid += 1
+
+        ## convert the edge metapath to the node metapath
+        edge2des = {}
+        edge2src = {}
+        for comb in self.g.canonical_etypes:
```

Comment:
```suggestion
        for src_type, etype, dst_type in self.g.canonical_etypes:
```
This might make the code more readable.